Run cargo check on udfs

jbeisen opened the pull request:
Add an API endpoint that accepts the contents of the UDFs editor and
sends it to the controller, and in turn to the compiler service, which
runs 'cargo check'. Pass any errors back to the API and show them in the
console. Also show the query validation errors in the Errors tab for
consistency.

Note that the UDFs are put into a new 'udfs' crate in the build
directory for the purpose of cargo check, and written again the
'pipeline' crate for the actual pipeline. A future commit will change
the pipeline to use the udfs crate.


---

Query error:

![image](https://github.com/ArroyoSystems/arroyo/assets/8881183/da849ed6-7f85-462c-b04e-60f456053071)


UDF error:

![image](https://github.com/ArroyoSystems/arroyo/assets/8881183/283d22c7-a480-44ac-8cc3-86502db6d571)

jbeisen commented on arroyo-controller/src/lib.rs:
```diff
@@ -427,6 +434,64 @@ impl ControllerGrpc for ControllerServer {
+        use std::time::SystemTime;
+        use std::str::FromStr;
+        use serde::{Deserialize, Serialize};
+    };
```
@mwylde I looked at the imports in `compile_pipeline_main` and kind of guessed what we want for udfs. What do you think?

jbeisen replied:
I've changed it so the exact contents of the UDFs editor are written to the build directory, so it's now up to the user to write whatever `use` statements they need.

mwylde commented on arroyo-api/src/pipelines.rs:
```diff
@@ -339,35 +348,82 @@ pub async fn validate_pipeline(
+        }
+    };
+
+    let pipeline_graph_validation_result = match compile_sql(&sql, &auth_data, &client).await {
```
Does it make sense to validate the sql first, so that we can report errors back more quickly?

jbeisen replied:
Yup, these are now separate endpoints.

mwylde commented on arroyo-api/src/pipelines.rs:
```diff
@@ -98,11 +108,7 @@ where
-        warn!("{:?}", err);
-        bad_request(format!("{}", err.root_cause()))
-    })?;
+    .with_context(|| "failed to generate SQL program")?;
```
this loses the detailed sql errors. 

For example, `select * from blah;` which on master tells me the actual error (blah doesn't exist) now just returns "failed to generate SQL program"

jbeisen replied:
Fixed

mwylde commented on arroyo-compiler-service/src/main.rs:
```diff
@@ -261,4 +263,54 @@ impl CompilerGrpc for CompileService {
+        request: Request<CheckUdfsModReq>,
+    ) -> Result<Response<CheckUdfsResp>, Status> {
+        // write udf to build_dir udfs module
+        let build_dir = &self.build_dir;
```
this could use some logging so it's clear what the compiler service is doing (and timing information is also helpful)

jbeisen replied:
Sure, I've added some logging. Are you suggesting something like timing how long `cargo check` takes?

mwylde replied:
Yes, like the existing timing information we have for compilation: https://github.com/ArroyoSystems/arroyo/blob/master/arroyo-compiler-service/src/main.rs#L208-L210

jbeisen replied:
Done

mwylde reviewed (changes requested):
This is a good start. I think it needs a couple more iterations before being ready to merge.

* We no longer get useful SQL error messages, which is a big degradation in the UX
* I think we can avoid slowing down SQL checking when there are no UDFs (probably the most common case)
* The UDF errors are often too vague to be useful (like "mismatched type" — which types? where?)
* This will need changes in the docker images so this can work outside of our dev environment — I would specifically build a docker-single image and make sure that works
* It might make sense to have udf checking be a separate API call from graph validation, which would allow caching of UDF results (that way you wouldn't need to recompile the udfs every time, even when they haven't changed)

jbeisen commented on arroyo-rpc/src/types.rs:
```diff
@@ -13,6 +13,27 @@ pub struct ValidatePipelinePost {
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct UdfsValidationResult {
+    pub udfs: Option<Vec<Udf>>,
```
The logic here is that `udfs` (and `PipelineGraphValidationResult::graph`) are only Some if the validation was successful and there are no errors, i.e. only one of these fields is Some. I originally had this as an enum like this:

```
pub enum UdfsValidationResult {
    Ok(Vec<Udf>),
    Error(Vec<String>),
}
```

but this creates a bit of a weird API json schema, since there isn't really a standard way of representing in json the Rust style of enums variants with values.

mwylde replied:
That still doesn't really make sense to me as an api consumer. Why are we returning the udfs at all?

mwylde replied:
I guess the weird thing is that we're using rust-style errors for an http endpoint, where you'd expect an error response in the error case.

mwylde replied:
Anyways I think this is fine for now but we should probably think about how to more naturally express this in REST for the next version

jbeisen replied:
> Why are we returning the udfs at all

My thinking is that POST endpoints are typically supposed to return a representation of the object that was created, and even though we're not creating an object with this request, we can still return the thing that was validated, kind of just like saying _this was the thing we validated_.

> does it make sense to distinguish between the empty vec case and none cases for errors

This is debatable but I think most people would [say](https://stackoverflow.com/questions/1969993/is-it-better-to-return-null-or-empty-collection) an empty vector is preferred because it avoids the class of null/undefined related errors. 

> I guess the weird thing is that we're using rust-style errors for an http endpoint, where you'd expect an error response in the error case.

I'm not sure that's the proper use of HTTP error codes, since the _errors_ we're talking about here originate from the code provided by the client, not on the server. I think 200 responses are still appropriate when the thing we're validating ends up being invalid.

jbeisen commented on arroyo-rpc/src/types.rs:
```diff
@@ -13,6 +13,27 @@ pub struct ValidatePipelinePost {
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct PipelineGraphValidationResult {
+    pub graph: Option<PipelineGraph>,
```
See above

jbeisen commented on arroyo-rpc/proto/rpc.proto:
```diff
@@ -425,6 +430,21 @@ message CompileQueryResp {
+
 service CompilerGrpc {
   rpc CompileQuery(CompileQueryReq) returns (CompileQueryResp);
+  rpc CheckUdfsMod(CheckUdfsModReq) returns (CheckUdfsResp);
```
The string that the controller sends to the compiler service is the pretty printed source code that goes in the udfs _module's_ lib.rs - 
```
prettyplease::unparse(&parse_quote! {
    #imports
    
    #(#udf_tokens )*
}
```

so I named it like this to differentiate it from the `CheckUdfsReq` that the api sends to the controller, which is just the udfs strings.

jbeisen replied:
I've renamed this to `CheckUdfs` since it now uses the same `CheckUdfsReq` as the controller.

jbeisen commented on arroyo-api/src/pipelines.rs:
```diff
@@ -339,35 +348,82 @@ pub async fn validate_pipeline(
+
+    let udf_validation_result = if check_udfs_resp.result == ValidationResult::Error as i32 {
+        UdfsValidationResult {
+            udfs: None,
```
Exactly one of these two fields will be Some, the other None, so it behaves like an enum. I think this format translates better to JSON for the api.

jbeisen commented on arroyo-api/src/pipelines.rs:
```diff
@@ -57,15 +60,24 @@ where
                 schema_provider
                     .add_rust_udf(&udf.definition)
-                    .map_err(|e| bad_request(format!("Could not process UDF: {:?}", e)))?;
+                    .map_err(|e| anyhow!(format!("Could not process UDF: {:?}", e)))?;
```
This function no longer uses those functions, but they're used elsewhere in the API, where I think they're useful because they give us some error message consistency.

mwylde replied:
I don't quite follow what you're saying. I mean, previously we were able to distinguish between user errors (like invalid udf language) and system errors (like we failed to talk to the database), and now those all get collapsed into anyhow::error.

jbeisen commented on arroyo-api/src/pipelines.rs:
```diff
@@ -339,35 +348,82 @@ pub async fn validate_pipeline(
-        })
+    let udf_strings = validate_pipeline_post
+        .udfs
+        .clone()
```
Removed.

jbeisen commented on arroyo-api/src/pipelines.rs:
```diff
@@ -339,35 +348,82 @@ pub async fn validate_pipeline(
+    // check the udfs
+    let mut controller = ControllerGrpcClient::connect(state.controller_addr.clone())
+        .await
+        .map_err(log_and_map)?;
```
Yup, makes sense. I've cleaned this so that the controller and compiler service both return appropriate `Status` responses and the API logs warnings.

jbeisen commented on arroyo-api/src/pipelines.rs:
```diff
@@ -339,35 +348,82 @@ pub async fn validate_pipeline(
-                value_type: edge.weight().value.to_string(),
-                edge_type: format!("{:?}", edge.weight().typ),
+
+    let check_udfs_resp = controller
```
Since it's now two separate endpoints, this endpoint will only be called if UDFs are defined.

jbeisen commented on build_dir/udfs/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
```
Done.

mwylde commented on build_dir/udfs/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+chrono = "0.4"
+serde = "1.0"
+serde_json = "1.0"
+arrow = { workspace = true}
```
This doesn't seem to have been done — I'm still seeing all of those dependencies in the Cargo.toml

mwylde commented on arroyo-api/src/pipelines.rs:
```diff
@@ -306,29 +313,27 @@ impl Into<Job> for DbPipelineJob {
 #[utoipa::path(
     post,
-    path = "/v1/pipelines/validate",
+    path = "/v1/pipelines/validate_query",
```
maybe leave this as `validate` to avoid unnecessary changes to the API?

jbeisen replied:
We're changing the response format, and to be consistent with the new endpoint I think we should rename it.

mwylde replied:
do we need to change the response format? since it's now doing the same thing as it was previously

jbeisen replied:
In my opinion it's cleanest to have both validate endpoints as similar as possible, and like I say [here](https://github.com/ArroyoSystems/arroyo/pull/359#discussion_r1357562372), I think this type of "error" that we're reporting with these endpoints is distinct from the type of "error" that HTTP response codes are supposed to represent. But if you feel strongly I can put it back.

mwylde commented on arroyo-api/src/pipelines.rs:
```diff
@@ -339,35 +344,112 @@ pub async fn validate_pipeline(
+    bearer_auth: BearerAuth,
+    WithRejection(Json(validate_udfs_post), _): WithRejection<Json<ValidateUdfsPost>, ApiError>,
+) -> Result<Json<UdfValidationResult>, ErrorResp> {
+    let _client = client(&state.pool).await?;
```
shuold this be removed?

mwylde replied:
that was supposed to be for the line above `_client`

jbeisen replied:
Sure

jbeisen replied:
Yes, done.

mwylde commented on arroyo-compiler-service/src/main.rs:
```diff
@@ -261,4 +270,70 @@ impl CompilerGrpc for CompileService {
+            .current_dir(&self.build_dir)
+            .arg("check")
+            .arg("--package=udfs")
+            .arg("--message-format=json")
```
ooh nice

mwylde commented on arroyo-console/package.json:
```diff
@@ -33,6 +33,7 @@
     "@types/react-syntax-highlighter": "^15.5.7",
     "ajv": "^8.12.0",
     "ajv-formats": "^2.1.1",
+    "ansi-to-react": "^6.1.6",
```
is this used?

jbeisen replied:
Unfortunately not, I've removed it.

mwylde reviewed (commented):
Overall looking good, a few more comments.

From the UX side, can we switch automatically to the error tab if there are validation failures? Currently the experience is a bit annoying because every time you click "check" it moves to the graph, so you constantly have to switch back to the errors tab as you're iterating.

Nit: the scrolling is a bit weird:
![image](https://github.com/ArroyoSystems/arroyo/assets/100920/95bb1a74-77e6-4e29-ab0e-0d9f6f07c8e7)

I'd expect the whole thing to scroll, rather than individual errors

jbeisen commented:
I've fixed the scroll issue and made it switch to the Errors tab when appropriate. I also moved the Alert into the tab - 

![image](https://github.com/ArroyoSystems/arroyo/assets/8881183/24d11ed4-8cf7-4a6c-ae01-d5ec2f8aaf97)
